src/fetch/mod.rs

```diff
@@ -56,3 +64,49 @@ pub async fn fetch_object_http<T: Clone, Kind: DeserializeOwned>(
 
     res.json_limited().await
 }
+
+/// Signed version of `fetch_object_http_signed`. This will sign the GET request
+/// using the private key of the given actor, which allows to implement secure
+/// federation mode.
+pub async fn fetch_object_http_signed<T: Clone, Kind: DeserializeOwned, A: Actor>(
+    url: &Url,
+    data: &Data<T>,
+    actor: &A,
+) -> Result<Kind, Error> {
+    let config = &data.config;
+    // dont fetch local objects this way
+    debug_assert!(url.domain() != Some(&config.domain));
+    config.verify_url_valid(url).await?;
+    info!("Fetching remote object {}", url.to_string());
+
+    let counter = data.request_counter.fetch_add(1, Ordering::SeqCst);
+    if counter > config.http_fetch_limit {
+        return Err(Error::RequestLimit);
+    }
+
+    let req = config
+        .client
+        .get(url.as_str())
+        .header("Accept", FEDERATION_CONTENT_TYPE)
+        .timeout(config.request_timeout);
+
+    let private_key_pem = actor
+        .private_key_pem()
+        .ok_or(anyhow!("Actor does not have a private key to sign with"))?;
+    let req = sign_request(
+        req,
+        actor.id(),
+        String::new(),
+        private_key_pem,
+        data.config.http_signature_compat,
+    )
+    .await?;
+
+    let res = config.client.execute(req).await.map_err(Error::other)?;
+
+    if res.status() == StatusCode::GONE {
+        return Err(Error::ObjectDeleted);
+    }
+
+    res.json_limited().await
+}
```

Comment:
You could merge this into `fetch_object_http()` by passing an `Option<Actor>`. Or to keep the API cleaner and avoid a breaking change, make a private method `fetch_object_http_internal()` with param `Option<Actor>` and have these two methods be wrappers for the internal one.